feat(reminders): review reminders RecyclerView

ericli3690 opened the pull request:
## Purpose / Description
- Added RecyclerView to ScheduleReminders fragment.
- Created ScheduleRemindersAdapter, which holds the state of each individual row in the RecyclerView. The adapter has functionality in the form of functions passed to it via its constructor to keep most of the key functionality centralized within ScheduleReminders and to avoid injecting a Context into the adapter, which should optimally be context-free. I tried to keep the adapter as short as possible; it only governs the UI of each individual reminder in the RecyclerView.
- Added `setDeckNameFromScopeView` method, which retrieves deck names from the collection for displaying in the UI. Deck names are cached to prevent repeated calls to the collection.
- Added empty placeholder methods for adding reminders and editing reminders that will be filled in by a future PR.
- Added logic for toggling whether reminders are enabled and saving that state.
- Added `onDeckSelected`, a method that will allow the code to detect if the AddEditReminderDialog (coming soon) has had the deck selected in the deck selection dialog changed. We must catch this in ScheduleReminders because this onChange event has to be caught by a Fragment, whereas the AddEditReminderDialog is just a dialog that gets briefly removed when the deck selection dialog appears. The data is forwarded to the AddEditReminderDialog via a FragmentResult, a very simple way of passing information between fragments.
- Added constants that will be used for pushing information to and from the AddEditReminderDialog to the ScheduleReminders companion object.
- Fixed a small Timber call to use format arguments rather than interpolation.

## Fixes
* For GSoC 2025: Review Reminders

## Approach
- I've tried to explain the motivation for implementation details in Purpose / Description. Let me know if anything is unclear.

## How Has This Been Tested?
- Launches using a physical Samsung S23, API 34.
- The full review reminders CRUD works when this PR is integrated with my AddEditReminderDialog and TimePicker code. Full testing should be possible once those are also merged.

## Checklist
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

<!--- Uncomment this section ONLY if this PR introduces new resources (external libraries, icons etc)
## Licenses
_For each new external resource, add a row to the table below:_

| Library | Description | License |
| --- | --- | --- |
| Sample Icon Library | Sample Description | [The Apache Software License, Version 2.0](http://www.apache.org/licenses/LICENSE-2.0.txt) |

**Maintainers:**

* [ ] Add the https://github.com/ankidroid/Anki-Android/labels/Licenses label
* [ ] Update the [licenses](https://github.com/ankidroid/Anki-Android/wiki/Licences) wiki when merging
--->

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ScheduleRemindersAdapter.kt:
```diff
@@ -0,0 +1,72 @@
+import com.ichi2.anki.R
+
+class ScheduleRemindersAdapter(
+    val reminders: MutableList<ReviewReminder>,
```
I considered this, but it felt like useless boilerplate to create a `sortReminders` or `addReminder` method when I could just... sort the reminders directly from ScheduleReminders. My personal preference is to keep the Adapter class as lean as possible and keep the actual non-trivial code within ScheduleReminders, but please let me know if I should prioritize making `reminders` private and I'll add some getters and setters.

ericli3690 replied:
No longer an issue, as we are using `ListAdapter` now, which allows me to explicitly manage the reminders directly from the `ScheduleReminders` class.

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ScheduleRemindersAdapter.kt:
```diff
@@ -0,0 +1,72 @@
+        holder.reminder = reminder
+
+        setDeckNameFromScopeForView(reminder.scope, holder.deckTextView)
+        holder.timeTextView.text = reminder.time.toString()
```
ReviewReminderTime has a custom toString method. Ignoring this suggestion.

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ScheduleReminders.kt:
```diff
@@ -71,12 +118,130 @@ class ScheduleReminders : Fragment(R.layout.fragment_schedule_reminders) {
+                is ReviewReminderScope.DeckSpecific -> database.getRemindersForDeck(scope.did)
+            }
+        val remindersAsList = remindersAsMap.values.sortedBy { it.time.toSecondsFromMidnight() }
+        adapter.reminders.addAll(remindersAsList)
+        adapter.notifyDataSetChanged()
```
Isn't that... basically the same thing?

I get that maybe notifying for a change in the item range might be a biiit more efficient, but looking through the AnkiDroid codebase, we mostly use `notifyDataSetChanged` over `notifyItemRangeChanged(0, items.size)`. Maintainers, would you like me to prioritize consistency with the rest of the codebase here or should I change this to `notifyItemRangeInserted`?

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ScheduleReminders.kt:
```diff
@@ -71,12 +118,130 @@ class ScheduleReminders : Fragment(R.layout.fragment_schedule_reminders) {
+        view: TextView,
+    ) {
+        when (scope) {
+            is ReviewReminderScope.Global -> view.text = "All Decks"
```
We plan on doing this at the end of August.

criticalAY commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ScheduleRemindersAdapter.kt:
```diff
@@ -0,0 +1,72 @@
+import com.google.android.material.materialswitch.MaterialSwitch
+import com.ichi2.anki.R
+
+class ScheduleRemindersAdapter(
```
You can use Diff callback here

ericli3690 replied:
Oh, I didn't know about this feature! Thanks for pointing it out, it streamlines a lot of my code. Implemented.

criticalAY reviewed (changes requested):
Small change rest is good

Arthur-Milchior reviewed (commented):
LGTM once @criticalAY request is considered

ericli3690 commented:
- Modified the ScheduleRemindersAdapter to be a ListAdapter rather than a RecyclerView.Adapter. Gave it a `diffCallback`.
- Moved RecyclerView `reminders` content state controlling data structure out of the adapter and into ScheduleReminders. Made it a HashMap for cleaner access operations without element-by-element search for deletion, etc.
- Modified the toggle listener to no longer use the adapter's provided `position` as an argument, as it sometimes caused out of index exceptions.
- Created a `triggerUIUpdates` method for ease of posting the reminders to the adapter.

criticalAY reviewed (approved):
Nice! thanks

ericli3690 commented:
Minor edit!

- Added a `catchDatabaseExceptions` wrapper method to ScheduleReminders. By wrapping calls to the database with this function, any database errors caught fail gracefully and show an error dialog to the user. If the database access takes too long, a progress bar shows up, too.
- Previously, this was a method that also included schema migration logic etc., but that's now been moved to https://github.com/ankidroid/Anki-Android/pull/18856, so it's now possible for me to move this small snippet here instead.

ericli3690 commented:
Minor edit!

- Non-functional change. Made some code in `toggleReminderEnabled` more readable, removing unnecessary null handling.
- Rebased.
